### PR TITLE
[MXNET-414] Tutorial on visualizing CNN decisions using Grad-CAM

### PR DESCRIPTION
**Pull request by indhub**
## Description ##
Tutorial on visualizing CNN decisions using Grad-CAM

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- [x] Code is well-documented: 
- [x] For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html

### CODE REVIEW

**Comment by ThomasDelteil** [comment]
@indhub I think you forgot to edit the docs/tutorials/index.md file to add your tutorial to the list

File: docs/tutorials/vision/cnn_visualization.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+We'll use pre-trained weights (trained on ImageNet) from model zoo instead of training the model from scratch.
+
+```python
+vgg_spec = {11: ([1, 1, 2, 2, 2], [64, 128, 256, 512, 512]),
+            13: ([2, 2, 2, 2, 2], [64, 128, 256, 512, 512]),
+            16: ([2, 2, 3, 3, 3], [64, 128, 256, 512, 512]),
+            19: ([2, 2, 4, 4, 4], [64, 128, 256, 512, 512])}
+
+def get_vgg(num_layers, pretrained=False, ctx=mx.cpu(),
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: It would be helpful to add some explanation to what is happening exactly here. 
Question: Isn't it possible to download the model from the model zoo using `.vgg16(pretrained=True)` and then on specific layers do `.collect_params()` and load these params on the custom model ?

File: docs/tutorials/vision/cnn_visualization.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+network = vgg16(pretrained=True, ctx=mx.cpu())
+```
+
+We define some helpers to read image files from disk and display multiple images in a row in Jupyter notebook.
+
+```python
+def read_image_mxnet(path):
+    with open(path, 'rb') as fp:
+        img_bytes = fp.read()
+    return mx.img.imdecode(img_bytes)
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: there is a `mx.image.imread()` that does this

File: docs/tutorials/vision/cnn_visualization.md [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+def read_image_cv(path):
+    return cv2.resize(cv2.cvtColor(cv2.imread(path), cv2.COLOR_BGR2RGB), image_sz)
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: I would suggest to simply use `mx.image.imread()` and then call `.asnumpy()` before plotting. There shouldn't be a need of a dependency on `cv2` that way.

File: docs/tutorials/vision/cnn_visualization.md [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+synset = []
+with open('synset.txt', 'r') as f:
+    synset = [l.rstrip().split(' ', 1)[1].split(',')[0] for l in f]
+    
+def get_class_name(cls_id):
+    return "%s (%d)" % (synset[cls_id], cls_id)
+
+def run_inference(net, data):
+    out = net(data)
+    return out.argmax(axis=1).asnumpy()[0].astype(int)
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: Suggestion: there is also a `topk()` function that is quite useful especially when getting several predictions and could be worth publicizing.

File: docs/tutorials/vision/cnn_visualization.md [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+Next, we'll write a method to get an image, preprocess it, predict category and visualize the prediction. We'll use `gradcam.visualize()` to create the visualizations. `gradcam.visualize` returns a tuple with the following visualizations:
+
+1. **Grad-CAM:** This is a heatmap superimposed on the input image showing which part(s) of the image contributed most to the CNN's decision.
+2. **Guided Grad-CAM:** Guided Grad-CAM shows which exact pixels contributed the most to the CNN's decision.
+3. **Saliency map:** Saliency map is a monochrome image showing which pixels contributed the most to the CNN's decision. Sometimes, it is easier to see the areas in the image that most influence the output in a monochrome image than in a color image.
+
+```python
+def visualize(net, img_path, conv_layer_name):
+    image = read_image_mxnet(img_path)
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: following previous comment, this could be `mx.image.imread(img_path)`

File: docs/tutorials/vision/cnn_visualization.md [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+    image = preprocess(image)
+    image = image.expand_dims(axis=0)
+    
+    pred_str = get_class_name(run_inference(net, image))
+    
+    orig_img = read_image_cv(img_path)
+    vizs = gradcam.visualize(net, image, orig_img, conv_layer_name)
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: following previous comment, you can get read of line 192, and replace `orig_img` with `image.asnumpy()`

File: docs/tutorials/vision/cnn_visualization.md [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+    return (pred_str, (orig_img, *vizs))
+```
+
+Next, we need to get the name of the last convolutional layer that extracts features from the image. We use the gradient information flowing into the last convolutional layer of the CNN to understand the importance of each neuron for a decision of interest. We are interested in the last convolutional layer because convolutional features naturally retain spatial information which is lost in fully connected layers. So, we expect the last convolutional layer to have the best compromise between high level semantics and detailed spacial information. The neurons in this layer look for semantic class specific information in the image (like object parts).
+
+In our network, feature extractors are added to a HybridSequential block named features. You can list the layers in that block by just printing `network.features`. You can see that the topmost convolutional layer is at index 28. `network.features[28]._name` will give the name of the layer.
+
+```python
+last_conv_layer_name = network.features[28]._name
+print(last_conv_layer_name)
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: it would be nice to have the output of this line in the tutorial `.md` you can add a `<!--notebook-skip-line-->` so that it doesn't show in the downloaded notebook.

File: docs/tutorials/vision/cnn_visualization.md [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+```python
+show_images(*visualize(network, "volcano.jpg", last_conv_layer_name))
+```
+
+<img src="https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/cnn_visualization/volcano.png" alt="Visualizing CNN decision"/><!--notebook-skip-line-->
+
+While it is not immediately evident why the network thinks this volcano is a spider, after looking at the Grad-CAM visualization, it is hard to look at the volcano and not see the spider!
+
+Being able to visualize why a CNN makes the decision it makes is a powerful tool to diagnose prediction failures. Even when the network is making correct predictions, visualization is a powerful tool to verify the network is making its decisions based on the right features and not some correlation which happens to exist in the training data.
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: Suggestion:
*Being able to visualize why a CNN makes the decision it makes *-> Being able to visualize why a CNN predicts specific classes is a...

*visualization is a powerful tool* -> visualizing activations is an important step to verify that..

File: docs/tutorials/vision/cnn_visualization.md [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+
+Visualization method demonstrated in this tutorial applies to a wide variety of network architectures and a wide variety of tasks beyond classification - like VQA and image captioning. Any type of differentiable output can be used to create the visualizations shown above. Visualization techniques like these solve (at least partially) the long standing problem of interpretability of neural networks.
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: The visualization method*

File: docs/tutorials/vision/cnn_visualization.md [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+show_images(*visualize(network, "hummingbird.jpg", last_conv_layer_name))
+```
+
+<img src="https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/cnn_visualization/hummingbird.png" alt="Visualizing CNN decision"/><!--notebook-skip-line-->
+
+```python
+show_images(*visualize(network, "jellyfish.jpg", last_conv_layer_name))
+```
+
+<img src="https://raw.githubusercontent.com/dmlc/web-data/master/mxnet/example/cnn_visualization/jellyfish.png" alt="Visualizing CNN decision"/><!--notebook-skip-line-->
```
Comment (ThomasDelteil) [inline docs/tutorials/vision/cnn_visualization.md]: Great visualizations!! Love them. However I find the heatmap difficult to read, what is the scale? What is hot what is cold 😄

**Comment by ThomasDelteil** [comment]
Awesome tutorial @indhub! Makes me wonder if the visualization could make its way to the contrib package at some point rather than example folder :)

File: docs/tutorials/vision/cnn_visualization.md [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+
+Visualizations also help build confidence about the predictions of a model. For example, even if a model correctly predicts birds as birds, we would want to confirm that the model bases its decision on the features of bird and not on the features of some other object that might occur together with birds in the dataset (like leaves).
+
+In this tutorial, we show how to visualize the predictions made by convolutional neural networks using Gradient-weighted Class Activation Mapping. Unlike many other visualization methods, Grad-CAM can be used on a wide variety of CNN model families - CNNs with fully connected layers, CNNs used for structural outputs (e.g. captioning), CNNs used in tasks with multi-model input (e.g. VQA) or reinforcement learning without architectural changes or re-training.
+
+In the rest of this notebook, we will explain how to visualize predictions made by [VGG-16](https://arxiv.org/abs/1409.1556). We begin by importing the required dependencies. `gradcam` module contains the implementation of visualization techniques used in this notebook.
```
[the remark this answers is not in the document]
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: VGG is used for a couple of reasons:
1. ResNet is too big to include in a notebook without spoiling easiness to read.
2. VGG16 the network used in the original paper.

File: docs/tutorials/vision/cnn_visualization.md [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+
+import mxnet as mx
+from mxnet import gluon
+
+from matplotlib import pyplot as plt
+import numpy as np
+import cv2
+
+gradcam_file = "gradcam.py" 
+base_url = "https://raw.githubusercontent.com/indhub/mxnet/cnnviz/example/cnn_visualization/{}?raw=true"
```
[the remark this answers is not in the document]
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: Will remember to switch.

File: docs/tutorials/vision/cnn_visualization.md (hunk 1 above)
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: I added more comments and simplified the code. Let me know if this still looks complicated.

File: docs/tutorials/vision/cnn_visualization.md (hunk 7 above)
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: Done.

File: docs/tutorials/vision/cnn_visualization.md (hunk 10 above)
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: Heatmap used is [COLORMAP_HSV](https://docs.opencv.org/2.4/modules/contrib/doc/facerec/colormaps.html). It has more colors in the range. But I get your point. Maybe something like COLORMAP_COOL is more intuitive.

File: docs/tutorials/vision/cnn_visualization.md (hunk 4 above)
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: `topk` is good but given this is a visualization tutorial, I would prefer keeping this simple. Single prediction creates more compact output images.

File: docs/tutorials/vision/cnn_visualization.md (hunk 10 above)
Comment (indhub) [reply docs/tutorials/vision/cnn_visualization.md]: Changed colormap to cv2. COLORMAP_COOL
